src/backend/distributed/planner/distributed_planner.c

```diff
@@ -304,44 +295,11 @@ distributed_planner(Query *parse,
 }
 
 
-/*
- * ErrorIfQueryHasUnsupportedMergeCommand walks over the query tree and bails out
- * if there is no Merge command (e.g., CMD_MERGE) in the query tree. For merge,
- * looks for all supported combinations, throws an exception if any violations
- * are seen.
- */
-static void
-ErrorIfQueryHasUnsupportedMergeCommand(Query *queryTree, List *rangeTableList)
-{
-	/*
-	 * Postgres currently doesn't support Merge queries inside subqueries and
-	 * ctes, but lets be defensive and do query tree walk anyway.
-	 *
-	 * We do not call this path for fast-path queries to avoid this additional
-	 * overhead.
-	 */
-	if (!ContainsMergeCommandWalker((Node *) queryTree))
-	{
-		/* No MERGE found */
-		return;
-	}
-
-
-	/*
-	 * In Citus we have limited support for MERGE, it's allowed
-	 * only if all the tables(target, source or any CTE) tables
-	 * are are local i.e. a combination of Citus local and Non-Citus
-	 * tables (regular Postgres tables).
-	 */
-	ErrorIfMergeHasUnsupportedTables(queryTree, rangeTableList);
-}
-
-
 /*
  * ContainsMergeCommandWalker walks over the node and finds if there are any
  * Merge command (e.g., CMD_MERGE) in the node.
  */
-static bool
+bool
 ContainsMergeCommandWalker(Node *node)
 {
```

Comment:
This is from original PG-15 porting code and I left it as-is without paying much attention.

`https://github.com/citusdata/citus/commit/3ed6fea1cf622bd5a69c425fcdef33a336686fc7`

Now I'm thinking it may be unnecessary as MERGE should be on top of the tree, but having said that, the current routine should never call query_tree_walker() as we yank out even before entering it? I'm fine with simplifying it, let me know either way.

Reply:
I guess if you enable these two tests, I'd be in favor of simplifying the code: https://github.com/citusdata/citus/blob/main/src/test/regress/expected/pg15.out#L429-L438


My main motivation is that reader of this function might think that Merge can be any place in the query, whereas it is only supported as top-level command

Reply:
OK, I've enabled these two tests (wasn't aware of their presence and context, good we have them) and removed the walker-routine.